keps/sig-auth/4412-projected-service-account-tokens-for-kubelet-image-credential-providers/README.md

```diff
@@ -0,0 +1,1034 @@
++       // +optional
++       ServiceAccountAnnotations []string
+```
+
+Example credential provider configuration:
+
+```yaml
+apiVersion: kubelet.config.k8s.io/v1
+kind: CredentialProviderConfig
+providers:
+  - name: acr-credential-provider
+    matchImages:
+      - "*.registry.io/*"
+    defaultCacheDuration: "10m"
+    apiVersion: credentialprovider.kubelet.k8s.io/v1
+    serviceAccountTokenAudience: my-audience
+    serviceAccountAnnotations:
+      - domain.io/identity-id
+      - domain.io/identity-type
+```
+
+##### Kubernetes API Server (KAS) Changes
+
+Add a new flag `--image-pull-credential-provider-audiences` to KAS to allow configuring the list of audiences for which the Kubelet is allowed to generate service account tokens for image pulls.
+
+The node audience restriction will be enforced by the [NodeRestriction admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction)
+and kubelet will only be allowed to generate a service account token for the audience configured in the credential provider configuration if it is present in the list of audiences configured in KAS.
+
+#### Credential Provider Request API
+
+Add `ServiceAccountToken` and `ServiceAccountAnnotations` fields to `CredentialProviderRequest`.
+
+1. `ServiceAccountToken` is the service account token bound to the pod for which the image is being pulled.
+    If the `ServiceAccountTokenAudience` field is configured in the kubelet's credential provider configuration, the token will be sent to the plugin.
+2. `ServiceAccountAnnotations` is a map of annotations on the KSA for which the image is being pulled. Only annotations defined in the `ServiceAccountAnnotations` field in the credential provider configuration will be passed to the plugin. If the `ServiceAccountAnnotations` field is not set in the configuration, this field will be empty.
+
+```diff
+--- a/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
++++ b/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider/types.go
+@@ -32,6 +32,16 @@ type CredentialProviderRequest struct {
+        // credential provider plugin request. Plugins may optionally parse the image
+        // to extract any information required to fetch credentials.
+        Image string
++
++       // serviceAccountToken is the service account token bound to the pod for which
++       // the image is being pulled. This token is only sent to the plugin if the
++       // audience field is configured in the kubelet's credential provider configuration.
++       ServiceAccountToken string
++
++       // serviceAccountAnnotations is a map of annotations on the service account bound to the
++       // pod for which the image is being pulled. The list of annotations in the service account
++       // that need to be passed to the plugin is configured in the kubelet's credential provider
++       // configuration.
++       ServiceAccountAnnotations map[string]string
```

Comment:
`ServiceAccountAnnotationKeys` in credential provider config will only be allowed to be set when the audience field is configured. This will be handled in validation, so we will not end up in a situation where the token is empty but annotations exist.